### PR TITLE
Implement micro-benchmarks via JMH

### PR DESCRIPTION
**Pull request by vlsi**
Here I add the micro-benchmarking module.
The purpose of the very first test (`StatementTest`) is to compare the performance of execution

``` sql
select name from hr.emps where empno=?
```

via four approaches:
1) plain java foreach looping through employees in `HrSchema` and finding the required `name`
2) `executeQuery(String)` with using literal for employee id
3) `prepareStatement(...); ps.setInt(...); ps.executeQuery()`
4) `ps.setInt(...); ps.executeQuery()` -- cached `PreparedStatement`

Here are the results:

```
Benchmark                                  Mode   Samples         Mean   Mean error    Units
n.h.o.StatementTest.forEach               thrpt        10   130285,170    31122,871   ops/ms
n.h.o.StatementTest.executeQuery          thrpt        10        0,043        0,007   ops/ms
n.h.o.StatementTest.prepareBindExecute    thrpt        10        0,042        0,008   ops/ms
n.h.o.StatementTest.bindExecute           thrpt        10      177,867        4,508   ops/ms
```

Note that cached and fully JIT'ed `PreparedStatement` looks to be 1000 times slower that foreach loop, however it achieves not that bad rate of 180 ops/ms == 180'000 / sec

A particular test can be investigated as follows: `java -Djmh.stack.lines=10 -jar ./target/ubenchmarks.jar ".*bindExecute.*" -prof hs_comp,hs_gc,stack -f 1 -wi 5 -i 20` (this runs `bindExecute` test with stack profiling in 1 jvm fork with 5 warmup iterations and 20 measurement iterations)

Here is sample result (it looks like lots of time is spent in creating of `DataContext` and in `OptiqConnectionProperty.wrap`):

```
       Stack |   5,7%   RUNNABLE java.util.Hashtable.hash
             |                   java.util.Hashtable.get
             |                   java.util.Properties.enumerateStringProperties
             |                   java.util.Properties.stringPropertyNames
             |                   net.hydromatic.avatica.ConnectionConfigImpl.parse
             |                   net.hydromatic.avatica.BuiltInConnectionProperty.wrap
             |                   net.hydromatic.avatica.ConnectionConfigImpl.timeZone
             |                   net.hydromatic.avatica.AvaticaConnection.getTimeZone
             |                   net.hydromatic.optiq.jdbc.OptiqConnectionImpl$DataContextImpl.<init>
             |                   net.hydromatic.optiq.jdbc.OptiqConnectionImpl.createDataContext
             | 
             |   5,7%   RUNNABLE java.util.Hashtable.<init>
             |                   java.util.Hashtable.<init>
             |                   java.util.Properties.stringPropertyNames
             |                   net.hydromatic.avatica.ConnectionConfigImpl.parse
             |                   net.hydromatic.optiq.config.OptiqConnectionProperty.wrap
             |                   net.hydromatic.optiq.jdbc.OptiqConnectionImpl$OptiqConnectionConfigImpl.autoTemp
             |                   net.hydromatic.optiq.jdbc.OptiqResultSet.execute
             |                   net.hydromatic.optiq.jdbc.OptiqResultSet.execute
             |                   net.hydromatic.avatica.AvaticaConnection.executeQueryInternal
             |                   net.hydromatic.avatica.AvaticaPreparedStatement.executeQuery
             | 
             |   5,7%   RUNNABLE sun.util.locale.LocaleUtils.isAlphaNumericString
             |                   sun.util.locale.UnicodeLocaleExtension.isKey
             |                   java.util.Locale.getUnicodeLocaleType
             |                   java.util.Calendar.createCalendar
             |                   java.util.Calendar.getInstance
             |                   net.hydromatic.avatica.AvaticaResultSet.<init>
             |                   net.hydromatic.optiq.jdbc.OptiqResultSet.<init>
             |                   net.hydromatic.optiq.jdbc.OptiqJdbc41Factory.newResultSet
             |                   net.hydromatic.optiq.jdbc.OptiqJdbc41Factory.newResultSet
             |                   net.hydromatic.avatica.AvaticaConnection.executeQueryInternal
             | 
             |   4,5%   RUNNABLE java.lang.String.toUpperCase
             |                   java.lang.String.toUpperCase
             |                   net.hydromatic.avatica.ConnectionConfigImpl.parse
             |                   net.hydromatic.optiq.config.OptiqConnectionProperty.wrap
             |                   net.hydromatic.optiq.jdbc.OptiqConnectionImpl$OptiqConnectionConfigImpl.autoTemp
             |                   net.hydromatic.optiq.jdbc.OptiqResultSet.execute
             |                   net.hydromatic.optiq.jdbc.OptiqResultSet.execute
             |                   net.hydromatic.avatica.AvaticaConnection.executeQueryInternal
             |                   net.hydromatic.avatica.AvaticaPreparedStatement.executeQuery
             |                   net.hydromatic.optiq.StatementTest.bindExecute
             | 
             |   4,5%   RUNNABLE java.util.HashMap.addEntry
             |                   java.util.LinkedHashMap.addEntry
             |                   java.util.HashMap.put
             |                   net.hydromatic.avatica.ConnectionConfigImpl.parse
             |                   net.hydromatic.optiq.config.OptiqConnectionProperty.wrap
             |                   net.hydromatic.optiq.jdbc.OptiqConnectionImpl$OptiqConnectionConfigImpl.autoTemp
             |                   net.hydromatic.optiq.jdbc.OptiqResultSet.execute
             |                   net.hydromatic.optiq.jdbc.OptiqResultSet.execute
             |                   net.hydromatic.avatica.AvaticaConnection.executeQueryInternal
             |                   net.hydromatic.avatica.AvaticaPreparedStatement.executeQuery
             | 
             |   4,5%   RUNNABLE java.lang.String.toUpperCase
             |                   java.lang.String.toUpperCase
             |                   net.hydromatic.avatica.ConnectionConfigImpl.parse
             |                   net.hydromatic.optiq.config.OptiqConnectionProperty.wrap
             |                   net.hydromatic.optiq.jdbc.OptiqConnectionImpl$OptiqConnectionConfigImpl.spark
             |                   net.hydromatic.optiq.jdbc.OptiqConnectionImpl.createDataContext
             |                   net.hydromatic.optiq.jdbc.MetaImpl.createCursor
             |                   net.hydromatic.avatica.AvaticaResultSet.execute
             |                   net.hydromatic.optiq.jdbc.OptiqResultSet.execute
             |                   net.hydromatic.optiq.jdbc.OptiqResultSet.execute
```

### CODE REVIEW

File: ubenchmark/pom.xml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  <parent>
+    <artifactId>optiq-parent</artifactId>
+    <groupId>net.hydromatic</groupId>
+    <version>0.7-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <properties>
+    <top.dir>${project.basedir}/..</top.dir>
+    <build.timestamp>${maven.build.timestamp}</build.timestamp>
```
Comment (julianhyde) [inline ubenchmark/pom.xml]: is build.timestamp needed? top project has it.

**Comment by julianhyde** [comment]
This looks like a useful change, but let's discuss where is the best place to put this.

One place would be in `core/test`. Dependencies on a third-party library, namely `org.openjdk.jmh`, wouldn't be a problem -- I don't mind adding dependencies to tests, in any module. But we couldn't dependencies on other modules, such as optiq-mongodb.

So, another place would be `plus/test`. I added that module recently as a place to add miscellaneous adapters. I also intended it as a place to add big system tests (that take a long time to run, use a lot of memory, or depend on a lot of third-party libraries).

`plus` is bit of a grab-bag. Benchmarking isn't quite testing, but it is very similar. I think that `StatementTest` would fit in well alongside `TpchTest`.

Can you make a case why `StatementTest` would be in a new module `ubenchmark`? I'm not against the idea - but I want to hear the justification.

File: ubenchmark/pom.xml (hunk 1 above)
Comment (vlsi) [inline ubenchmark/pom.xml]: I've copied both properties verbatim from optiq-core.
Have no idea if `build.timestamp` is required (why if `maven.build.timestamp is available`?).

`top.dir` is definitely required for checkstyle, thus I've added both properties.

**Comment by vlsi** [comment]
1) Typical build for ubenchmark includes maven-shade-plugin (i.e. repackaging of all the dependencies), thus I am not sure if we really want that for all the `plus` scope
2) The use-case for ubenchmark is quite different from tpch/other tests: in ubenchmarks you run small number of tests lots of times. It is more like a research than a regular test, so you run with different versions of optiq, different jvm options, etc. The test itself is not big in terms of memory or time, but the number of required invocations to get to stable performance is rather high.
3) For jmh benchmarks there is a well-known entry point and command line. Thus if we want incorporating jmh and tpch into a single module, we'll need to rewrite all test execution to jmh conventions (that is not bad, but it does require time).

**Comment by julianhyde** [comment]
Why build.timestamp is required: http://stackoverflow.com/questions/13228472/how-to-acces-maven-build-timestamp-for-resource-filtering

**Comment by julianhyde** [comment]
Those are good reasons to put micro-benchmarking in its own module.

Merged as https://github.com/julianhyde/optiq/commit/7fdc1615f36bb9d59bfaf3679b861cebbff1d412.

**Pull request closed without merging**